perf: use getAllReportErrors from a derived value

TMisiukiewicz opened the pull request:
<!-- If necessary, assign reviewers that know the area or changes well. Feel free to tag any additional reviewers you see fit. -->

### Explanation of Change
<!-- Explain what your change does and how it addresses the linked issue -->
This PR aims to replace occurencies of `getAllReportErrors` to read a value from a derived value instead.

Comparison with `main`:

**Action:** Send a message
**Environment:** Web, 4x CPU throttling
**Function:** `getAllReportErrors`
**Before:** 86ms
**After:** 1.2ms

### Fixed Issues
<!---
1. Please postfix `$` with a URL link to the GitHub issue this Pull Request is fixing. For example, `$ https://github.com/Expensify/App/issues/<issueID>`.
2. Please postfix  `PROPOSAL:` with a URL link to your GitHub comment, which contains the approved proposal (i.e. the proposal that was approved by Expensify).  For example, `PROPOSAL: https://github.com/Expensify/App/issues/<issueID>#issuecomment-1369752925`

Do NOT add the special GH keywords like `fixed` etc, we have our own process of managing the flow.
It MUST be an entire link to the github issue and your comment proposal ; otherwise, the linking and its automation will not work as expected.

Make sure this section looks similar to this (you can link multiple issues using the same formatting, just add a new line):

$ https://github.com/Expensify/App/issues/<issueID>
$ https://github.com/Expensify/App/issues/<issueID(comment)>

Do NOT only link the issue number like this: $ #<issueID>
--->
$ https://github.com/Expensify/App/issues/59350
PROPOSAL:


<!--- 
If you want to trigger adhoc build of hybrid app from specific Mobile-Expensify PR please link it like this:

MOBILE-EXPENSIFY: https://github.com/Expensify/Mobile-Expensify/pull/<PR-number>

--->

### Tests
<!---
Add a numbered list of manual tests you performed that validates your changes work on all platforms, and that there are no regressions present.
Add any additional test steps if test steps are unique to a particular platform.
Manual test steps should be written so that your reviewer can repeat and verify one or more expected outcomes in the development environment.

For example:
1. Click on the text input to bring it into focus
2. Upload an image via copy paste
3. Verify a modal appears displaying a preview of that image
--->
1. Open the app
2. Verify that your RBR statuses are displayed in LHN
3. Open any chat that does not have RBR status
4. Split expense and upload an image that is not a correct receipt so the scan will fail
5. Verify the RBR appeared in LHN
- [x] Verify that no errors appear in the JS console

### Offline tests
<!---
Add any relevant steps that validate your changes work as expected in a variety of network states e.g. "offline", "spotty connection", "slow internet", etc. Manual test steps should be written so that your reviewer and QA testers can repeat and verify one or more expected outcomes. If you are unsure how the behavior should work ask for advice in the `#expensify-open-source` Slack channel.
--->
n/a
### QA Steps
<!---
Add a numbered list of manual tests that can be performed by our QA engineers on the staging environment to validate that your changes work on all platforms, and that there are no regressions present.
Add any additional QA steps if test steps are unique to a particular platform.
Manual test steps should be written so that the QA engineer can repeat and verify one or more expected outcomes in the staging environment.

For example:
1. Click on the text input to bring it into focus
2. Upload an image via copy paste
3. Verify a modal appears displaying a preview of that image

It's acceptable to write "Same as tests" if the QA team is able to run the tests in the above "Tests" section.
--->
// TODO: These must be filled out, or the issue title must include "[No QA]."
Same as tests
- [x] Verify that no errors appear in the JS console

### PR Author Checklist
<!--
This is a checklist for PR authors. Please make sure to complete all tasks and check them off once you do, or else your PR will not be merged!
-->

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [x] MacOS: Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is either coming verbatim from figma or has been approved by marketing (in order to get marketing approval, ask the Bug Zero team member to add the Waiting for copy label to the issue)
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If the PR modifies the UI (e.g. new buttons, new UI components, changing the padding/spacing/sizing, moving components, etc) or modifies the form input styles:
    - [x] I verified that all the inputs inside a form are aligned with each other.
    - [x] I added `Design` label and/or tagged `@Expensify/design` so the design team can review the changes.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] I added [unit tests](https://github.com/Expensify/App/blob/main/tests/README.md) for any new feature or bug fix in this PR to help automatically prevent regressions in this user flow.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>

https://github.com/user-attachments/assets/d406f793-7176-4fe8-8d3a-981459f7f3db


<!-- add screenshots or videos here -->

</details>

<details>
<summary>Android: mWeb Chrome</summary>

<!-- add screenshots or videos here -->

</details>

<details>
<summary>iOS: Native</summary>

https://github.com/user-attachments/assets/ddfb0360-76bd-4d0f-a46b-4a85e2f18c8f


<!-- add screenshots or videos here -->

</details>

<details>
<summary>iOS: mWeb Safari</summary>

https://github.com/user-attachments/assets/c44dcae6-083c-4df9-be9d-cdaed2b53c1b


<!-- add screenshots or videos here -->

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

<!-- add screenshots or videos here -->

https://github.com/user-attachments/assets/d9ed4c9c-0306-4275-9a37-e3dfabb928a3


</details>

<details>
<summary>MacOS: Desktop</summary>

https://github.com/user-attachments/assets/0ff36343-6500-4bd6-ac6d-870b0a61ed35


<!-- add screenshots or videos here -->

</details>

TMisiukiewicz committed:
expose value from a derived value

TMisiukiewicz committed:
replace in SidebarUtils

TMisiukiewicz committed:
replace in OptionListUtils

TMisiukiewicz committed:
replace occurencies of getAllReportErrors

TMisiukiewicz committed:
fix CI jobs

TMisiukiewicz commented:
Waiting for a merge of https://github.com/Expensify/App/pull/61376

TMisiukiewicz committed:
fix lint job

TMisiukiewicz committed:
fix lint changed in DebugReportPage

TMisiukiewicz committed:
Merge remote-tracking branch 'upstream/main' into perf/use-getAllReportErrors-from-derived-value

TMisiukiewicz committed:
reduce diff

TMisiukiewicz committed:
update submodule

TMisiukiewicz committed:
fix changed files

TMisiukiewicz committed:
fix CI jobs

TMisiukiewicz committed:
Merge branch 'main' into perf/use-getAllReportErrors-from-derived-value

TMisiukiewicz committed:
Merge branch 'main' into perf/use-getAllReportErrors-from-derived-value

TMisiukiewicz committed:
use reportAttributes from param

TMisiukiewicz committed:
update submodule

TMisiukiewicz committed:
remove deprecated getTransaction from SidebarUtils

TMisiukiewicz commented:
@hoangzinh could you take a look when you have a moment please? 🙏

hoangzinh reviewed (commented):
All goods, just 3 small feedbacks

TMisiukiewicz commented on src/libs/SidebarUtils.ts:
```diff
@@ -165,6 +164,18 @@ Onyx.connect({
+        }
+        allTransactions = Object.fromEntries(Object.entries(value).filter(([, transaction]) => !!transaction));
+    },
+});
```
`getTransaction` is deprecated now, linter is forcing to change it 😢

TMisiukiewicz committed:
Merge remote-tracking branch 'upstream/main' into perf/use-getAllReportErrors-from-derived-value

TMisiukiewicz committed:
remove hasReportErrors

TMisiukiewicz committed:
migrate from deprecated getTransaction

TMisiukiewicz commented:
@hoangzinh I addressed your comments, additionally I resolved this comment from another PR: https://github.com/Expensify/App/pull/62464#discussion_r2100467719 here as lint changed job is forcing to do the same changes in both PRs

TMisiukiewicz committed:
update deps in DebugReportPage

TMisiukiewicz commented:
can anyone with permissions rerun the CI jobs? Something went wrong here

TMisiukiewicz committed:
update submodule

TMisiukiewicz committed:
add getAllReportErrors tests

TMisiukiewicz committed:
update submodule

TMisiukiewicz commented on src/libs/actions/OnyxDerived/configs/reportAttributes.ts:
```diff
@@ -132,6 +132,7 @@ export default createOnyxDerivedValueConfig({
                 isEmpty: generateIsEmptyReport(report),
                 brickRoadStatus,
                 requiresAttention,
+                reportErrors,
```
@hoangzinh I've just added a couple of tests 👍

mountiny commented:
We should not, the submodule chance should not be here @TMisiukiewicz also new conflicts please 🙌

TMisiukiewicz committed:
Merge remote-tracking branch 'upstream/main' into perf/use-getAllReportErrors-from-derived-value

TMisiukiewicz commented:
@hoangzinh @mountiny done ✅

TMisiukiewicz committed:
Merge remote-tracking branch 'upstream/main' into perf/use-getAllReportErrors-from-derived-value

TMisiukiewicz committed:
Merge remote-tracking branch 'upstream/main' into perf/use-getAllReportErrors-from-derived-value

TMisiukiewicz commented:
@hoangzinh all jobs passed now after merging the latest main 👍

hoangzinh reviewed (approved):
LGTM

mountiny reviewed (dismissed):
Thanks @TMisiukiewicz changes look good to me and thanks @hoangzinh for a thorough review and testing.

This change has a high chance of introducing some deploy blockers so please if they come up after the next deploy @TMisiukiewicz make it your top priority and monitor new blockers as they come in. Thank you!

mjasikowski reviewed (dismissed):
Looks gucci, no objections here

TMisiukiewicz committed:
Merge remote-tracking branch 'upstream/main' into perf/use-getAllReportErrors-from-derived-value

mountiny reviewed (approved):
Thanks!

OSBotify commented:
:hand: This PR was not deployed to staging yet because QA is ongoing. It will be automatically deployed to staging after the next production release.
